add-privileged-users-in-room

GoodGuyMarco opened the pull request:
Jira-Ticket: https://element-io.atlassian.net/browse/PSB-169
Related PR: https://github.com/matrix-org/matrix-js-sdk/pull/2892.
---
Adds a new section under "Room Settings" > "Roles & Permissions" which adds the possibility to multiselect users from this room and grant them more permissions.

![OSS_1](https://user-images.githubusercontent.com/20201936/202673835-d5fe05c3-632e-4e91-86e3-07eee8bb066b.PNG)

When searching users via the new `AutocompleteInput` component, suggestions are shown and can be selected. Select a power level and hit the "Apply" button to confirm your changes. Users should appear in the "Privileged Users" section afterwards.

![OSS_2](https://user-images.githubusercontent.com/20201936/202673894-b7735d17-196e-4a14-b3c7-b8143b23d079.PNG)

## Checklist

* [x] Tests written for new code (and old code if feasible)
* [x] Linter and other CI checks pass
* [x] Sign-off given on the changes (see [CONTRIBUTING.md](https://github.com/matrix-org/matrix-react-sdk/blob/develop/CONTRIBUTING.md))

Notes: Adds a new section under "Room Settings" > "Roles & Permissions" which adds the possibility to multiselect users from this room and grant them more permissions.

element-web notes: Adds a new section under "Room Settings" > "Roles & Permissions" which adds the possibility to multiselect users from this room and grant them more permissions.

Signed-off-by: Marco Bartelt marco.bartelt.ext@bwi.de

<!-- CHANGELOG_PREVIEW_START -->
---
Here's what your changelog entry will look like:

## ✨ Features
 * Adds a new section under "Room Settings" > "Roles & Permissions" which adds the possibility to multiselect users from this room and grant them more permissions. ([\#9596](https://github.com/matrix-org/matrix-react-sdk/pull/9596)). Contributed by @GoodGuyMarco.<!-- CHANGELOG_PREVIEW_END -->

t3chguy commented:
@jakewb-b could you get a designed tasked with reviewing it?

jakewb-b commented:
> @jakewb-b could you get a designer tasked with reviewing it?

Yup - already been flagged to @gaelledel who was also involved in the original discussions on this.

gsouquet reviewed (changes requested):
Thank you for opening this pull request!
I've added a few comments that would need to be adressed before this can be merged in. There's also a few issue in regards to TypeScript strict mode that have been picked up by our CI that would need your attention.

I have not yet checked the tests you added, but will do once the rest of the code reaches it's more final shape

GoodGuyMarco commented on res/css/structures/_AutocompleteInput.pcss:
```diff
@@ -0,0 +1,118 @@
+.mx_AutocompleteInput {
+    position: relative;
+
+    &_matches {
```
Fixed in this commit https://github.com/matrix-org/matrix-react-sdk/pull/9596/commits/01220e9a653b216944d01f43fd9eb539944fae96

GoodGuyMarco commented on res/css/structures/_AutocompleteInput.pcss:
```diff
@@ -0,0 +1,118 @@
+        }
+
+        &_description {
+            color: grey;
```
I have replaced that by `$secondary-content` 🧑‍🎨

GoodGuyMarco commented on res/css/structures/_AutocompleteInput.pcss:
```diff
@@ -0,0 +1,118 @@
+    &_suggestion {
+        display: flex;
+        align-items: center;
+        padding: 8px;
```
I tried replacing all hardcoded instances with the spacing variables. Point me out if I missed any 🕵️

GoodGuyMarco commented on res/css/structures/_AutocompleteInput.pcss:
```diff
@@ -0,0 +1,118 @@
+                font-weight: normal !important;
+                font-size: 1.4rem;
+            }
+        }
```
Can you clarify what you mean by that?

gsouquet replied:
Having `!important` is usually an anti-pattern and this code is using a lot of it... which is slightly worrying and might be a sign that we're abusing another component in fitting these requirements

GoodGuyMarco replied:
I've investigated the styling and found that global input styles are interfering with these styles. I've tried to get rid of some `!important`'s but did not manage to do that without breaking component encapsulation. The input styles in this component are kind of special as the input needs to be hidden. 

Maybe we could use a different approach where we make the actual input visible for the user, position the selection pills absolute above the input and dynamically increase the padding on the left, but I don't know if that is worth the effort to get rid of some `!important`'s.

GoodGuyMarco commented on res/css/structures/_AutocompleteInput.pcss:
```diff
@@ -0,0 +1,118 @@
+                height: 24px;
+                padding-left: 8px;
+                padding-right: 8px;
+                color: #ffffff;
```
I agree. Could you point me to a combination of background and font color variables that I should use for this? 

![pill](https://user-images.githubusercontent.com/20201936/203041169-11baa0c1-f61f-44e8-9517-1e3b9e4a5b1e.PNG)

gsouquet replied:
The @matrix-org/design review should point you in the right direction

GoodGuyMarco commented on src/i18n/strings/de_DE.json:
```diff
@@ -3658,5 +3658,9 @@
+    "Add privileged users": "Privilegierte Benutzer hinzufügen",
+    "Give one or multiple users in this room more privileges": "Hier kannst du Raummitgliedern mehr Privilegien zuweisen.",
+    "Search users in this room ...": "Benutzer suchen ...",
+    "Apply": "Übernehmen"
```
Didn't know that. Removed the german translations ✅

GoodGuyMarco commented on src/components/views/settings/AddPrivilegedUsers.tsx:
```diff
@@ -0,0 +1,89 @@
+            <AccessibleButton
+                kind='primary'
+                disabled={!selectedUsers.length || isLoading}
+                onClick={onSubmit}
```
There was no `form` element until now. I've introduced one, but that leads to not beeing able to use the `SettingsFieldSet` component as it does not work well when beeing nested inside a `form`. I've reused the classes of the component but am unsure if this is a good solution. Any ideas?

GoodGuyMarco replied:
EDIT: Found an alternative solution using flexbox which is able to reuse the `SettingsFieldSet` component.

GoodGuyMarco commented on src/components/views/settings/AddPrivilegedUsers.tsx:
```diff
@@ -0,0 +1,89 @@
+        >
+            <AutocompleteInput
+                provider={userProvider.current}
+                placeholder={_t("Search users in this room ...")}
```
Done in https://github.com/matrix-org/matrix-react-sdk/pull/9596/commits/e3d2113d5d094ff396d4f3d44ffb4dbd7b25779a

GoodGuyMarco commented on src/components/structures/AutocompleteInput.tsx:
```diff
@@ -0,0 +1,242 @@
+    const [_selection, setSelection] = useState<ICompletion[]>([]);
+    const [suggestions, setSuggestions] = useState<ICompletion[]>([]);
+    const [query, setQuery] = useState<string>('');
+    const [focusedElement, setFocusedElement] = useState<HTMLElement>(null);
```
Tried that but apparrently the suggestions don't close when clicking away from the component 🤔

GoodGuyMarco replied:
I've changed the behaviour to work with a boolean `isFocused`.

GoodGuyMarco commented on src/components/structures/AutocompleteInput.tsx:
```diff
@@ -0,0 +1,242 @@
+    }, [query, maxSuggestions, provider, additionalFilter]);
+
+    useEffect(() => {
+        getSuggestions();
```
Yeah that's odd indeed (code was not written by me initially). I've changed the component to be controlled for `selection` and `onSelectionChange` as this makes it far more easy in my opinion.

GoodGuyMarco commented on src/components/structures/AutocompleteInput.tsx:
```diff
@@ -0,0 +1,242 @@
+            newSelection.splice(idx, 1);
+            if (onSelectionChange) {
+                onSelectionChange(newSelection);
+            }
```
I agree, but now it is kind of obsolet because of https://github.com/matrix-org/matrix-react-sdk/pull/9596#discussion_r1028956386

GoodGuyMarco commented:
Hey @gsouquet, thank you for your constructive feedback! I tried to address all of your comments. Some of them were fixable, others need your feedback.  

Could you help me regarding the TypeScript strict mode issue? Yesterday it showed me only errors in my changed files. Today it gives me 600 > errors, which is confusing. Maybe I could solve it with a rebase, but I'm not sure. Anyway it is not really convenient that I am not able to run this check locally like the Linters because that heavily increases the time spend on the feedback loop. Any ideas?

gsouquet commented on src/components/structures/AutocompleteInput.tsx:
```diff
@@ -0,0 +1,219 @@
+                    value={query}
+                    autoComplete="off"
+                    placeholder={hasPlaceholder() ? placeholder : null}
+                    onFocus={() => setFocused(true)}
+                    onBlur={() => setFocused(false)}
```
We actually have a react hook called `useFocus` that you could use here.

GoodGuyMarco replied:
Neat, thanks for the hint 👍

gsouquet reviewed (changes requested):
Thanks for adressing the feedback, a few minor things here. But this is definitely on the right track!
Next step would be to get a review from design.

A rebase with `develop` should help with CI going a bit crazy with the TS issues. If you want to test this locally, you can enable `strict: true` in `tsconfig.json` temporarily

GoodGuyMarco commented on src/components/structures/AutocompleteInput.tsx:
```diff
@@ -0,0 +1,219 @@
+        }
+    };
+
+    const _renderSuggestion = (completion: ICompletion): ReactNode => {
```
I agree and changed it as suggested!

GoodGuyMarco commented on res/css/structures/_AutocompleteInput.pcss:
```diff
@@ -0,0 +1,104 @@
+.mx_AutocompleteInput {
```
Upsi 👀

GoodGuyMarco commented:
@gsouquet Hey, I've just rebased against `develop` and addressed your second feedback loop last week. Could you have another look and ask for the design feedback?

gaelledel reviewed (approved):
looks good to me thanks!

gsouquet reviewed (approved):
Code looks good to me, there are still some design outstanding question for @gaelledel before we can merge this.

GoodGuyMarco commented:
Thank you for your feedback @gaelledel. I have a few questions, as speaking about visuals always leaves room for misunderstanding.

> 1. For the search users field: We need to add a proper search field component with the Search icon and the following string "Search users"

1. By _"...search field component with the search icon..."_ you mean that the input should have a magnifier icon on the left side (like in the screenshot)?

![search](https://user-images.githubusercontent.com/20201936/204763810-5937e660-fef2-402a-abe7-9d0feae77ba1.PNG)

2. By _"...and the following string "Search users"..."_ you want to replace the placeholder I suggested `Search users in this room...` by `Search users`, correct?

---

> 2. Search users field: As soon as the user clicks in the field, we should display results in the drop down. Currently it doesn't show anything

Currently it displays results (matching your search criteria) as soon as you start typing. Why do we think that it should appear right away and what are your expectations if we change that regarding following questions: 

1. How many results should appear when the room has many users. Should there be a maximum and what should it be?
2. Which users should appear first when there is no search criteria?

---

> 3. Search field select colour should be set to Quarternary

Do you mean the background color of a selected suggestion?

![suggestion](https://user-images.githubusercontent.com/20201936/204768615-6f3f175d-0234-4ad2-9971-3907ed940d1d.PNG)

---

> 4. When selecting "custom level" and clicking away from the fields it doesn't deselect the option which as a user does not allow me to rectify my choice to change a permission. No matter what I do, the Custom level is set to changeable numbers and does not give me the ability to change to another permission type

Do you mean the the user is "trapped" inside the "custom level" input and can't go back to the select field? This is actually the current behaviour of this component. I stumbled across this aswell, but would suggest not changing this in my PR as it seems unrelated and deserves another issue.

GoodGuyMarco commented:
@gaelledel 

### 🧑‍🎨 Added a magnifier on the left
---
![OSS_i1](https://user-images.githubusercontent.com/20201936/205290510-ac91c680-ea7e-4ef1-8f79-cb5c9d203002.PNG)

### 🧑‍🎨 Aligned look and feel to regular input components (focus border color, equal height)
---
![OSS_i2](https://user-images.githubusercontent.com/20201936/205290512-562392a0-feac-4edc-b21d-25da79e6ff7a.PNG)

### 🧑‍🎨 Updated hover and selected background color of the suggestions
---
![OSS_i3](https://user-images.githubusercontent.com/20201936/205290517-e4d01956-8e21-4b41-a015-687012b04798.PNG)

gaelledel reviewed (approved):
Reviewed with Marco and all good from me

gsouquet reviewed (approved):
Looks good to me, but the CI is not too happy. Let's fix that, and we should be able to merge this change in!

t3chguy commented:
Branch matching is likely kicking in, their js-sdk needs updating or deleting

GoodGuyMarco commented:
@t3chguy I just rebased the PR. What do you mean by "branch matching"?

t3chguy commented:
Our CI attempts to pick matching branches between all the various layers (js-sdk, react-sdk, element-web, etc) by finding branches of the same name in your forks. https://github.com/matrix-org/matrix-js-sdk/pull/2892 is such a branch, and that is based on an older version of develop so the branch needs either updating or deleting.

GoodGuyMarco commented:
@t3chguy Okay. I could not find the merged branch in matrix-js-sdk. Did you delete it already? If so, could you manually re-run the CI to see if that helped?

t3chguy commented:
We can't delete branches on your fork

GoodGuyMarco commented:
@t3chguy my bad, deleted the branch in our fork ✅

weeman1337 commented on test/components/structures/AutocompleteInput-test.tsx:
```diff
@@ -0,0 +1,244 @@
+        getCompletions: jest.fn().mockImplementation(async () => data),
+    }) as unknown as AutocompleteProvider;
+
+    beforeEach(async () => {
```
```suggestion
    beforeEach(() => {
```

weeman1337 commented on test/components/structures/AutocompleteInput-test.tsx:
```diff
@@ -0,0 +1,244 @@
+    });
+
+    const getEditorInput = () => {
+        const input = screen.getByTestId('autocomplete-input');
```
`getBy` already ensures that this element exists and explodes otherwise.
If you want to test, that some element exist you could do

```
expect(screen.queryByLabelText("something")).toBeInTheDocument();
```

weeman1337 commented on src/components/structures/AutocompleteInput.tsx:
```diff
@@ -0,0 +1,248 @@
+                    value={query}
+                    autoComplete="off"
+                    placeholder={hasPlaceholder() ? placeholder : undefined}
+                    data-testid="autocomplete-input"
```
We should better test by using labels / texts.

weeman1337 commented on test/components/structures/AutocompleteInput-test.tsx:
```diff
@@ -0,0 +1,244 @@
+
+        act(() => {
+            fireEvent.focus(input);
+            fireEvent.change(input, { target: { value: 'user' } });
```
We should use `type` here. Like mentioned in other places we should stay close on how users are acting. Typing into the field would be an option https://testing-library.com/docs/ecosystem-user-event/#typeelement-text-options

weeman1337 commented on test/components/structures/AutocompleteInput-test.tsx:
```diff
@@ -0,0 +1,244 @@
+            fireEvent.change(input, { target: { value: 'user' } });
+        });
+
+        await waitFor(() => expect(mockProvider.getCompletions).toHaveBeenCalledTimes(1));
```
We should check for the expected texts to be visible here.

weeman1337 replied:
Can be improved later

weeman1337 commented on test/components/structures/AutocompleteInput-test.tsx:
```diff
@@ -0,0 +1,244 @@
+        );
+
+        const editor = screen.getByTestId('autocomplete-editor');
+        const selection = within(editor).getAllByTestId("autocomplete-selection-item", { exact: false });
```
Same as in other places, we should better check for texts, labels etc..

weeman1337 commented on test/components/structures/AutocompleteInput-test.tsx:
```diff
@@ -0,0 +1,244 @@
+        return input;
+    };
+
+    it('should render suggestions when a query is set', async () => {
```
```suggestion
    it('should render suggestions when a query is set', () => {
```

weeman1337 commented on test/components/structures/AutocompleteInput-test.tsx:
```diff
@@ -0,0 +1,244 @@
+        expect(screen.getByTestId('autocomplete-matches').childNodes).toHaveLength(mockCompletion.length);
+    });
+
+    it('should render selected items passed in via props', async () => {
```
```suggestion
    it('should render selected items passed in via props', () => {
```

weeman1337 commented on test/components/structures/AutocompleteInput-test.tsx:
```diff
@@ -0,0 +1,244 @@
+        expect(selection).toHaveLength(mockCompletion.length);
+    });
+
+    it('should call onSelectionChange() when an item is removed from selection', async () => {
```
```suggestion
    it('should call onSelectionChange() when an item is removed from selection', () => {
```

weeman1337 commented on test/components/structures/AutocompleteInput-test.tsx:
```diff
@@ -0,0 +1,244 @@
+        expect(onSelectionChangeMock).toHaveBeenCalledWith([mockCompletion[1]]);
+    });
+
+    it('should render custom selection element when renderSelection() is defined', async () => {
```
```suggestion
    it('should render custom selection element when renderSelection() is defined', () => {
```

weeman1337 commented on test/components/views/settings/AddPrivilegedUsers-test.tsx:
```diff
@@ -0,0 +1,143 @@
+            fireEvent.mouseDown(matchOne);
+        });
+
+        act(() => {
```
Can this be merged to the previous `act`?

GoodGuyMarco replied:
Actually started with a merged `act` and divided it because it failed 🤔

weeman1337 replied:
Ok, lets keep it for now

weeman1337 reviewed (commented):
Thanks for adding the tests. Tests should be more focused to mimic real user interaction like looking for texts / labels and typing / clicking instead of triggering change events.
